src/Simulation/EntryPointDriver/Validation.cs

```diff
@@ -1,117 +1,155 @@
-﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+﻿// Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
 
 using System;
 using System.Collections.Generic;
-using System.Diagnostics.CodeAnalysis;
+using System.Collections.Immutable;
 using System.Linq;
 
 namespace Microsoft.Quantum.EntryPointDriver
 {
     /// <summary>
-    /// Represents either a success or a failure of a process.
+    /// Represents either success with a value of type <typeparamref name="TValue"/>, or failure with an error of type
+    /// <typeparamref name="TError"/>.
     /// </summary>
-    /// <typeparam name="T">The type of the success value.</typeparam>
-    internal readonly struct Validation<T>
+    /// <typeparam name="TValue">The type of the value.</typeparam>
+    /// <typeparam name="TError">The type of the error.</typeparam>
+    internal abstract class Validation<TValue, TError>
```

Comment:
No, I don't think it was called `Either`, but it was called `Result` before. That was confusing with the Q# `Result` type which was also used. If it's called `Either`, then the cases would be `Left` and `Right`, which I think is more confusing than `Success` / `Failure` or `Ok` / `Error`.

The name comes from Scalaz's [`Validation`](https://javadoc.io/doc/org.scalaz/scalaz-core_2.12/latest/scalaz/Validation.html), which has similar behavior of accumulating errors in the `Sequence` method rather than stopping at the first error like `Either` would.